### PR TITLE
Support for `as` operator and user-defined `As` impls.

### PR DESCRIPTION
**Pull request by zygoloid**

### CODE REVIEW

File: explorer/testdata/as/implicitas.carbon [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// RUN: %{explorer} --parser_debug --trace_file=- %s 2>&1 | \
+// RUN:   %{FileCheck} --match-full-lines --allow-unused-prefixes %s
+// AUTOUPDATE: %{explorer} %s
+// CHECK: result: 5
+
+package ExplorerTest api;
+
+class A { var n: i32; }
+
+external impl A as ImplicitAs(i32) {
```
Comment (jonmeow) [inline explorer/testdata/as/implicitas.carbon]: rename file to implicit_as to represent the snake_case version of CameCase?

File: explorer/interpreter/type_checker.cpp [hunk 2]
```diff
@@ -2048,6 +2048,26 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
...
+              ExpectIsConcreteType(op.arguments()[1]->source_loc(), type));
+          ErrorOr<Nonnull<Expression*>> converted =
+              BuildBuiltinMethodCall(impl_scope, op.arguments()[0],
+                                     BuiltinInterfaceName{Builtins::As, type},
+                                     BuiltinMethodCall{"Convert"});
+          if (!converted.ok()) {
+            // We couldn't find a matching `impl`.
+            return CompilationError(e->source_loc())
+                   << "type error in `as`: "
+                   << "`" << *ts[0] << "` is not explicitly convertible to `"
```
Comment (jonmeow) [inline explorer/interpreter/type_checker.cpp]:
```suggestion
                   << "type error in `as`: `" << *ts[0] << "` is not explicitly convertible to `"
```

I wouldn't have put the backtic as a single-char string, you don't do it in the second example of this on the line. (and obviously this is going to need reformatting)

File: explorer/interpreter/type_checker.cpp [hunk 3]
```diff
@@ -2048,6 +2048,26 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
...
+                   << *type << "`:\n" << converted.error().message();
```
Comment (jonmeow) [inline explorer/interpreter/type_checker.cpp]: In the actual example, I missed that these were associated (:) and thought it was just two errors being printed. Not sure if there's a way to improve this by cropping out the duplicate source location or similar, though.

**Review by jonmeow — APPROVED** [review]
Looks good, just a couple minor comments that don't feel like I need to review responses to before you merge.

File: explorer/interpreter/type_checker.cpp (hunk 3 above)
Comment (zygoloid) [reply explorer/interpreter/type_checker.cpp]: I'll send a separate PR for that.

**Pull request merged**